### PR TITLE
ubuntu20.04 compilation issues

### PR DESCRIPTION
**Pull request by birbalsain**
#### Describe Bug or Feature
Getting warnings while compiling on ubuntu20 that is using higher gcc version 9.3. Read and write size doesn't match in snprintf(), structure sockaddr doesn't give correct size of ipv6. reading data from null string.
[ubuntu_issue.txt](https://github.com/openpbs/openpbs/files/5481183/ubuntu_issue.txt)

#### Describe Your Change
Uses structure sockaddr_storage to store ipv6. Provides required size to the writing destination. Uses if condition to check null string,

#### Attach Test and Valgrind Logs/Output

[ubuntu_build.txt](https://github.com/openpbs/openpbs/files/5483651/ubuntu_build.txt)

Description: Tests from given build on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, SLES12, SLES15
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,SLES12,SLES15
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|4663|3845|14|0|0|0|3831|

### CODE REVIEW

File: src/tools/pbs_idled.c [hunk 1]
```diff
@@ -117,7 +117,7 @@ main(int argc, char *argv[], char *envp[])
 	char *username;
 	struct xy cur_xy, prev_xy;
 	struct stat st;
-	char errbuf[256];
+	char errbuf[1037];
```
Comment (hirenvadalia) [inline src/tools/pbs_idled.c:120]: Can't we use some macro here instead this magic number?

File: src/tools/pbs_python.c [hunk 2]
```diff
@@ -1648,7 +1648,7 @@ main(int argc, char *argv[], char *envp[])
 		int	print_env = 0;
 		char	*tmp_str = NULL;
 		char	perf_label[MAXBUF];
-		char	perf_action[MAXBUFLEN];
+		char	perf_action[MAXBUFLEN + 13];
```
Comment (hirenvadalia) [inline src/tools/pbs_python.c:1651]: Please add comment saying we need `+ 13`

File: src/resmom/mom_hook_func.c [hunk 3]
```diff
@@ -750,7 +750,7 @@ run_hook(hook *phook, unsigned int event_type, mom_hook_input_t *hook_input,
 	pid_t myseq; /* just some unique sequence number */
 	char logmask[BUFSIZ];
 	char path_hooks_rescdef[MAXPATHLEN + 1];
-	char cmdline[2 * BUFSIZ + 1];
+	char cmdline[3 * BUFSIZ + 1];
```
Comment (hirenvadalia) [inline src/resmom/mom_hook_func.c:753]: better to add comment saying why `3 * size + 1`

File: src/tools/pbs_python.c (hunk 2 above)
Comment (birbalsain) [reply src/tools/pbs_python.c:1651]: perf_action is used to write hook perf output and description string, where hook perf output size is MAXBUFLEN so need additional size for description string.

Comment (hirenvadalia) [reply src/tools/pbs_python.c:1651]: @birbalsain Can you please add whatever you said in code comment?

File: src/cmds/qsub.c [hunk 4]
```diff
@@ -1833,7 +1833,7 @@ read_job_script(char *script)
 	int errflg; /* error code from get_script() */
 	struct stat statbuf;
 	char *bnp;
-	char basename[PBS_MAXJOBNAME + 1]; /* base name of script for job name*/
+	char basename[MAXPATHLEN + 1]; /* base name of script for job name*/
```
Comment (agrawalravi90) [inline src/cmds/qsub.c:1836]: why did you change this? PBS_MAXJOBNAME is supposed to be used for this exact purpose.

File: src/lib/Libtpp/tpp_util.c [hunk 5]
```diff
@@ -2081,18 +2081,19 @@ tpp_get_addresses(char *names, int *count)
 tpp_addr_t *
 tpp_get_local_host(int sock)
 {
-	struct sockaddr addr;
+	struct sockaddr_storage addrs;
+	struct sockaddr *addr = (struct sockaddr *)&addrs;
```
Comment (agrawalravi90) [inline src/lib/Libtpp/tpp_util.c:2085]: is this really ok? it's usually very dangerous to cast pointer of one type to another, unless you are casting a void pointer, It can lead to unexpected runtime issues.

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/cmds/qsub.c (hunk 4 above)
Comment (birbalsain) [reply src/cmds/qsub.c:1836]: basename holds script file that is static char script[MAXPATHLEN + 1] = "";

Comment (agrawalravi90) [reply src/cmds/qsub.c:1836]: then we should change PBS_MAXJOBNAME to be MAXPATHLEN + \<N>.

File: src/lib/Libtpp/tpp_util.c (hunk 5 above)
Comment (vstumpf) [reply src/lib/Libtpp/tpp_util.c:2085]: It should be fine, according to [spec](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html)

> When a sockaddr_storage structure is cast as a sockaddr structure, the ss_family field of the sockaddr_storage structure shall map onto the sa_family field of the sockaddr structure. When a sockaddr_storage structure is cast as a protocol-specific address structure, the ss_family field shall map onto a field of that structure that is of type sa_family_t and that identifies the protocol's address family.

Comment (agrawalravi90) [reply src/lib/Libtpp/tpp_util.c:2085]: Thanks Vincent. From what I understand, this states that it's ok to cast sockaddr_storage to sockaddr. My concern is regarding the pointer cast from sockaddr_storage * to sockaddr *. I've been bitten in the past by casting int * to long *, the compiler did weird things with it at higher optimization levels. So, just to be on the safe side, can we do "struct sockaddr addr = (struct sockaddr ) addrs;" instead ?

File: src/include/pbs_ifl.h [hunk 6]
```diff
@@ -466,7 +466,7 @@ enum mgr_obj {
 #define PBS_MAXPWLEN		256		/* max password length */
 #define PBS_MAXGRPN		256		/* max group name length */
 #define PBS_MAXQUEUENAME	15		/* max queue name length */
-#define PBS_MAXJOBNAME  	230		/* max job name length */
+#define PBS_MAXJOBNAME  	MAXPATHLEN + 1	/* max job name length */
```
Comment (hirenvadalia) [inline src/include/pbs_ifl.h:469]: This is not right... PBS_MAXJOBNAME is what we can pass in `-N` in qsub/pbs_rsub, not complete path to script. see `decode_jobname()` action func... Please revert this.

**Review by hirenvadalia — CHANGES_REQUESTED** [review]

File: src/cmds/qsub.c (hunk 4 above)
Comment (hirenvadalia) [reply src/cmds/qsub.c]: @agrawalravi90 PBS_MAXJOBNAME is not full path of job script, its max job name (`-N` in qsub/pbs_rsub) which is derived from job script's path if `-N` not given. So I don't agree with you that we should change PBS_MAXJOBNAME.

File: src/lib/Libtpp/tpp_util.c (hunk 5 above)
Comment (hirenvadalia) [reply src/lib/Libtpp/tpp_util.c:2085]: @agrawalravi90 Its ok to cast pointer also in this case, if you read spec link given by @vstumpf then it says that:
> The <sys/socket.h> header shall define the sockaddr_storage structure. This structure shall be:
>
>    Large enough to accommodate all supported protocol-specific address structures
>
>    Aligned at an appropriate boundary so that pointers to it can be cast as pointers to protocol-specific address structures and used to access the fields of those structures without alignment problems

File: src/cmds/qsub.c (hunk 4 above)
Comment (agrawalravi90) [reply src/cmds/qsub.c:1836]: If job name is derived from job script's path then does it not make sense that the job name buffer be at least MAXPATHLEN in size? Anyways, the variable name and comment here seemed to suggest that it should use buffer size for job name, but I'm ok with anything as long as compiler clears it.

File: src/lib/Libtpp/tpp_util.c (hunk 5 above)
Comment (agrawalravi90) [reply src/lib/Libtpp/tpp_util.c:2085]: Thanks for looking into it, I guess if the spec guarantees the safety then we can use it.

**Review by agrawalravi90 — APPROVED** [review]
Approving this since it seems urgent and I have no problems with the changes.

**Review by hirenvadalia — APPROVED** [review]
